src/hera/shared/_type_util.py

```diff
@@ -0,0 +1,78 @@
+"""Module that handles types and annotations."""
+
+from typing import Iterable, Optional, Tuple, Type, TypeVar, Union, cast, overload
+
+try:
+    from types import UnionType  # type: ignore
+except ImportError:
+    UnionType = Union  # type: ignore
+
+try:
+    from typing import Annotated, get_args, get_origin  # type: ignore
+except ImportError:
+    from typing_extensions import Annotated, get_args, get_origin  # type: ignore
+
+
+def is_annotated(annotation: type):
+    """Check annotation has Annotated type or not."""
+    return get_origin(annotation) is Annotated
+
+
+def unwrap_annotation(annotation: type) -> type:
+    """If the given annotation is of type Annotated, return the underlying type, otherwise return the annotation."""
+    if is_annotated(annotation):
+        return get_args(annotation)[0]
+    return annotation
+
+
+T = TypeVar("T")
+V = TypeVar("V")
+
+
+@overload
+def get_annotated_metadata(_: type, __: Type[T]) -> Optional[T]: ...
+
+
+@overload
+def get_annotated_metadata(_: type, __: Tuple[Type[T], Type[V]]) -> Optional[Union[T, V]]: ...
+
+
+# FIXME: Currently, mypy cannot guess following type hint properly: https://github.com/python/mypy/issues/17700
+#        def get_annotated_metadata(_: Any, __: Union[Type[T], Tuple[Type[T], ...]]) -> Optional[T]: ...
+#        Once fixed, remove overloads and add simpler type hints.
+def get_annotated_metadata(annotation, type_):
+    """If given annotation has metadata typed type_, return the metadata."""
+    args = get_args(annotation)
+    for arg in args[1:]:
+        if isinstance(type_, Iterable):
+            for t in type_:
+                if isinstance(arg, t):
+                    return arg
+        else:
+            if isinstance(arg, type_):
+                return arg
+    return None
+
+
+def get_origin_or_builtin(t: type) -> type:
+    """Get the unsubscripted version of t or the type of t itself if it's a built it, and cast it as a Python `type`.
+
+    This can be helpful if you want to use t with isinstance, issubclass, etc.,
+    """
+    if origin_type := get_origin(t):
+        return cast(type, origin_type)
+    return t
```

Comment:
some types are not the instance of type class. especially union.

```py
>>> from typing import Union, get_origin
>>> isinstance(get_origin(Union[int, str]), type)
False
>>> isinstance(Union, type)
False
```

Reply:
Good point...

Reply:
(Though we could still simplify the code to `return cast(type, get_origin(t) or t)`)

Reply:
If we're passing unions into this function then it seems as if it should be annotated as `(Any) -> Any`, since e.g. `int | str` is not considered a `type` by type-checkers, and neither is `UnionType`.

Reply:
In this case the signature of the function should probably use `Any` (as we are expecting `Union`s to possibly appear in `origin_type_issubclass` below).

Reply:
I checked related codes and just found that we don't need cast(type, ...) anymore. also, any -> any is right. fixed it!